File: core/src/main/java/art/arcane/iris/core/safeguard/IrisSafeguard.java
package art.arcane.iris.core.safeguard;

import art.arcane.iris.Iris;
import art.arcane.iris.core.safeguard.task.Diagnostic;
import art.arcane.iris.core.safeguard.task.Task;
import art.arcane.iris.core.safeguard.task.Tasks;
import art.arcane.iris.core.safeguard.task.ValueWithDiagnostics;
import art.arcane.iris.util.common.format.C;
import art.arcane.iris.util.common.scheduling.J;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class IrisSafeguard {
    private static volatile boolean forceShutdown = false;
    private static Map<Task, ValueWithDiagnostics<Mode>> results = Collections.emptyMap();
    private static Map<String, String> context = Collections.emptyMap();
    private static Map<String, List<String>> attachment = Collections.emptyMap();
    private static Mode mode = Mode.STABLE;
    private static int count = 0;

    private IrisSafeguard() {
    }

    public static void execute() {
        List<Task> tasks = Tasks.getTasks();
        LinkedHashMap<Task, ValueWithDiagnostics<Mode>> resultValues = new LinkedHashMap<>(tasks.size());
        LinkedHashMap<String, String> contextValues = new LinkedHashMap<>(tasks.size());
        LinkedHashMap<String, List<String>> attachmentValues = new LinkedHashMap<>(tasks.size());
        Mode currentMode = Mode.STABLE;
        int issueCount = 0;

        for (Task task : tasks) {
            ValueWithDiagnostics<Mode> result;
            try {
                result = task.run();
            } catch (Throwable e) {
                Iris.reportError(e);
                result = new ValueWithDiagnostics<>(
                        Mode.WARNING,
                        new Diagnostic(Diagnostic.Logger.ERROR, "Error while running task " + task.getId(), e)
                );
            }

            currentMode = currentMode.highest(result.getValue());
            resultValues.put(task, result);
            contextValues.put(task.getId(), result.getValue().getId());

            List<String> lines = new ArrayList<>();
            for (Diagnostic diagnostic : result.getDiagnostics()) {
                String[] split = diagnostic.toString().split("\\n");
                Collections.addAll(lines, split);
            }
            attachmentValues.put(task.getId(), lines);

            if (result.getValue() != Mode.STABLE) {
                issueCount++;
            }
        }

        results = Collections.unmodifiableMap(resultValues);
        context = Collections.unmodifiableMap(contextValues);
        attachment = Collections.unmodifiableMap(attachmentValues);
        mode = currentMode;
        count = issueCount;
    }

    public static Mode mode() {
        return mode;
    }

    public static Map<String, String> asContext() {
        return context;
    }

    public static Map<String, List<String>> asAttachment() {
        return attachment;
    }

    public static void splash() {
        Iris.instance.splash();
        printReports();
        printFooter();
    }

    public static void printReports() {
        switch (mode) {
            case STABLE -> Iris.info(C.BLUE + "0 Conflicts found");
            case WARNING -> Iris.warn(C.GOLD + "%s Issues found", count);
            case UNSTABLE -> Iris.error(C.DARK_RED + "%s Issues found", count);
        }

        for (ValueWithDiagnostics<Mode> value : results.values()) {
            value.log(true, true);
        }
    }

    public static void printFooter() {
        switch (mode) {
            case STABLE -> Iris.info(C.BLUE + "Iris is running Stable");
            case WARNING -> warning();
            case UNSTABLE -> unstable();
        }
    }

    public static boolean isForceShutdown() {
        return forceShutdown;
    }

    private static void warning() {
        Iris.warn(C.GOLD + "Iris is running in Warning Mode");
        Iris.warn(C.GRAY + "Some startup checks need attention. Review the messages above for tuning suggestions.");
        Iris.warn(C.GRAY + "Iris will continue startup normally.");
        Iris.warn("");
    }

    private static void unstable() {
        Iris.error(C.DARK_RED + "Iris is running in Danger Mode");
        Iris.error("");
        Iris.error(C.DARK_GRAY + "--==<" + C.RED + " IMPORTANT " + C.DARK_GRAY + ">==--");
        Iris.error("Critical startup checks failed. Iris will continue startup in 10 seconds.");
        Iris.error("Review and resolve the errors above as soon as possible.");
        J.sleep(10000L);
        Iris.info("");
    }
}

File: core/src/main/kotlin/art/arcane/iris/core/safeguard/IrisSafeguard.kt
package art.arcane.iris.core.safeguard

import art.arcane.iris.Iris
import art.arcane.iris.core.safeguard.task.Diagnostic
import art.arcane.iris.core.safeguard.task.Task
import art.arcane.iris.core.safeguard.task.ValueWithDiagnostics
import art.arcane.iris.core.safeguard.task.tasks
import art.arcane.iris.util.format.C
import art.arcane.iris.util.scheduling.J
import java.util.*

object IrisSafeguard {
    @Volatile
    private var forceShutdown = false
    private var results: Map<Task, ValueWithDiagnostics<Mode>> = emptyMap()
    private var context: Map<String, String> = emptyMap()
    private var attachment: Map<String, List<String>> = emptyMap()
    private var mode = Mode.STABLE
    private var count = 0

    @JvmStatic
    fun execute() {
        val results = LinkedHashMap<Task, ValueWithDiagnostics<Mode>>(tasks.size)
        val context = LinkedHashMap<String, String>(tasks.size)
        val attachment = LinkedHashMap<String, List<String>>(tasks.size)
        var mode = Mode.STABLE
        var count = 0
        for (task in tasks) {
            var result: ValueWithDiagnostics<Mode>
            try {
                result = task.run()
            } catch (e: Throwable) {
                Iris.reportError(e)
                result = ValueWithDiagnostics(
                    Mode.WARNING,
                    Diagnostic(Diagnostic.Logger.ERROR, "Error while running task ${task.id}", e)
                )
            }
            mode = mode.highest(result.value)
            results[task] = result
            context[task.id] = result.value.id
            attachment[task.id] = result.diagnostics.flatMap { it.toString().split('\n') }
            if (result.value != Mode.STABLE) count++
        }

        this.results = Collections.unmodifiableMap(results)
        this.context = Collections.unmodifiableMap(context)
        this.attachment = Collections.unmodifiableMap(attachment)
        this.mode = mode
        this.count = count
    }

    @JvmStatic
    fun mode() = mode

    @JvmStatic
    fun asContext() = context

    @JvmStatic
    fun asAttachment() = attachment

    @JvmStatic
    fun splash() {
        Iris.instance.splash()
        printReports()
        printFooter()
    }

    @JvmStatic
    fun printReports() {
        when (mode) {
            Mode.STABLE -> Iris.info(C.BLUE.toString() + "0 Conflicts found")
            Mode.WARNING -> Iris.warn(C.GOLD.toString() + "%s Issues found", count)
            Mode.UNSTABLE -> Iris.error(C.DARK_RED.toString() + "%s Issues found", count)
        }

        results.values.forEach { it.log(withStackTrace = true) }
    }

    @JvmStatic
    fun printFooter() {
        when (mode) {
            Mode.STABLE -> Iris.info(C.BLUE.toString() + "Iris is running Stable")
            Mode.WARNING -> warning()
            Mode.UNSTABLE -> unstable()
        }
    }

    @JvmStatic
    fun isForceShutdown() = forceShutdown

    private fun warning() {
        Iris.warn(C.GOLD.toString() + "Iris is running in Warning Mode")
        Iris.warn(C.GRAY.toString() + "Some startup checks need attention. Review the messages above for tuning suggestions.")
        Iris.warn(C.GRAY.toString() + "Iris will continue startup normally.")
        Iris.warn("")
    }

    private fun unstable() {
        Iris.error(C.DARK_RED.toString() + "Iris is running in Danger Mode")
        Iris.error("")
        Iris.error(C.DARK_GRAY.toString() + "--==<" + C.RED + " IMPORTANT " + C.DARK_GRAY + ">==--")
        Iris.error("Critical startup checks failed. Iris will continue startup in 10 seconds.")
        Iris.error("Review and resolve the errors above as soon as possible.")
        J.sleep(10000L)
        Iris.info("")
    }
}
